---
title: "Installation"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Installation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

You can install DEUS from GitHub with:

```{r gh-installation, eval=FALSE}
if (!require("devtools")) install.packages("devtools", repos='http://cran.us.r-project.org')
devtools::install_github("timjeske/DEUS", build_opts = c("--no-resave-data", "--no-manual"))
```

Alternatively, you can install DEUS from a local copy.

First, you need to unzip it:

```bash
unzip DEUS-master.zip
```

Then you can install it in R:

```{r, eval=FALSE}
if (!require("devtools")) install.packages("devtools", repos='http://cran.us.r-project.org')
devtools::install_local("./DEUS-master", build_opts = c("--no-resave-data", "--no-manual"))
```

### Prerequisites

The following system libraries are required:

* libssl-dev (Debian)/ openssl-devel (Fedora)
* libcurl4-openssl-dev (Debian)/ libcurl-devel (Fedora)
* libxml2-dev (Debian)/ libxml2-devel (Fedora)

If these system dependencies are fulfilled, all required R packages should be installed automatically.

Further, the DEUS pipeline requires the following tools:

* [CD-Hit](http://weizhongli-lab.org/cd-hit/) (cd-hit-est binary)
* [BLAST](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/) (blastn binary)
* Custom BLAST database used for annotation step


### Preparing a BLAST database
During the annotation step, a BLAST database is used to determine possible biological sources of each unique sequence. Depending on the use-case, two options are available how an appropriate database can be obtained:

1. Use existing databases available at e.g. [NCBI](ftp://ftp.ncbi.nlm.nih.gov/blast/db/). Database sets may also be retrieved automatically with `update_blastdb.pl`, which is part of the BLAST+ suite.
2. Build a custom database using BLAST `makeblastdb` and some fasta sequences:

```bash
cat DASHR.fa Homo_sapiens.GRCh37.75.cds.all.fa  Homo_sapiens.GRCh37.75.ncrna.fa > HumanDB.fa
makeblastdb -in HumanDB.fa -dbtype nucl -parse_seqids
```

### Preparing a condition file
In order to execute the differential expression analysis, a file specifying different conditions is necessary. 
Each condition file has to contain **sample names**, the respective **FASTQ files** as well as the matching **condition**. The file should be in **TSV format** (Tab-Separated Values).
```
sampleName      sample  condition
cond1_s1        cond1_s1.fwd.fq.gz      cond1
cond1_s2        cond1_s2.fwd.fq.gz      cond1
cond2_s1        cond2_s1.fwd.fq.gz      cond2
cond2_s2        cond2_s2.fwd.fq.gz      cond2
```

### Run example 
To test your setup, you can run our test script that will perform DEUS on our simulated example data. It can also be used as a template for further analyses. 
You can find the [test script](https://github.com/timjeske/DEUS/blob/master/sample_usage.R) in the [DEUS GitHub repository](https://github.com/timjeske/DEUS).
An additional [sample script](https://github.com/timjeske/DEUS/blob/master/sample_usage_withCl.R) performing an extended analysis using sequence clusters is also provided in the repository.
The individual steps of both scripts are explained in [Run DEUS](SampleUsage.html).
	
```r
Rscript sample_usage.R
```
