---
title: "Run DEUS"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Run DEUS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

### (Optional) preprocessing of FASTQ files

Before applying DEUS we strongly recommend to trim your input FASTQ files. 
[Trim Galore](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) can be used to do adapter and quality trimming. The `--max_length` parameter should be set to (read length - 1) to remove non-small RNA sequences.

```bash
# remove adapters (automatically recognized by Trim Galore)
/software/trim_galore_v0.x.x/trim_galore -q 0 --length 16 --max_length 49 $in_file -o $out_dir
# quality trimming (use dummy adapter to suppress adapter trimming)
/software/trim_galore_v0.x.x/trim_galore -q 20 -a CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC --stringency 50 $in_file -o $out_dir
```

### Installing DEUS

```{r, eval=FALSE}
if (!require("devtools")) install.packages("devtools", repos='http://cran.us.r-project.org')
devtools::install_github("timjeske/DEUS")
library(DEUS)

```

### Setting parameters

Before running the individual steps of the DEUS pipeline it is helpful to set all required parameters.
The `in_dir` is the folder where the (trimmed) FASTQ files are stored. The `phenofile` is the condition file for differential expression analysis (DEA) ([examplary condition file](DEUS.html#preparing-a-condition-file)).
The `out_dir` is the folder where DEA plots, clustering and summary output files will be stored.
To use BLAST, it is necessary to set the [BLAST binary](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/), your [BLAST database](DEUS.html#preparing-a-blast-database) and the number of threads to be used. 
To cluster resulting sequencing, the [cd-hit-est](http://weizhongli-lab.org/cd-hit/) binary is reuqired.

```{r, eval=FALSE}

in_dir <- "/data/fastq-files/"
phenofile <- "/data/condition.tsv"

out_dir <- "data/results"

blast_exec <- "/software/blast/bin/blastn"
blast_db <- "/data/My_Blast_DB.fa"
blast_ncores <- 2

cd_hit <- "/software/cdhit/cd-hit-est"
```

### Counting unique sequences

To create the count table for differential expression analysis, [createCountTableFromFastQs](../reference/createCountTableFromFastQs.html) counts all unique sequences in each FASTQ file given in your `phenofile`. To remove low expressed sequences and reduce the size of the count table [filterLowExp](../reference/filterLowExp.html) is applied. Sequences are kept if both groups of samples have a default minimum average sequence count of 10 or if one of both groups has zero counts for all samples. It can be useful to store the filtered count table in our output folder as counting takes some time.
```{r, eval=FALSE}
phenoInfo <- read.table(phenofile, header=T, row.names=1, check.names=FALSE)
countTable <- createCountTableFromFastQs(in_dir, phenoInfo=phenoInfo)
countTable <- filterLowExp(countTable, phenoInfo)
write.table(countTable, paste(out_dir,"AllCounts_filtered.tsv",sep="/"), col.names=T, quote=F, sep="\t", row.names=T)
```

### Running differential expression analysis

Differential expression analysis is done by calling the function [runDESeq2](../reference/runDESeq2.html). 
After filtering for significant sequences, a map is created by [createMap](../reference/createMap.html) which assigns a unique identifier to each sequence.
The results of [runDESeq2](../reference/runDESeq2.html) are additionally used as input for [getConditionCountStats](../reference/getConditionCountStats.html) to compute the mean and the standard deviation of the normalized counts for each analyzed conditions.

```{r, eval=FALSE}
design <- ~ condition
deResults <- runDESeq2(countTable, phenoInfo, design, map, out_dir)
sigResults <- deResults$deResult
sigResults <- sigResults[!is.na(sigResults$IHWPval) & sigResults$IHWPval < 0.05,]
map <- createMap(sigResults)

# get count stats
countStats <- getConditionCountStats(deResults$normCounts, phenoInfo)
```

### Running BLAST

To annotate significant sequences they need to transformed to a vector of sequences in FASTA style by using [sequencesAsFasta](../reference/sequencesAsFasta.html). Afterwards BLAST is executed by [runBlast](../reference/runBlast.html).
```{r, eval=FALSE}
sigSeqFasta <- sequencesAsFasta(sigResults,map)
blastResult <- runBlast(blast_exec, blast_db, blast_ncores, sigSeqFasta)
```

### Run clustering

The CD-Hit algorithm is executed via [runClustering](../reference/runClustering.html).
```{r, eval=FALSE}
  clustResult <- runClustering(cd_hit, sigSeqFasta, out_dir, 0.9, 0.9, 9, map)
```

### Merging the results

Finally, the group-wise mean and standard deviation of the normalized counts, the results of DEA, BLASTing and clustering are merged together by [mergeResults](../reference/mergeResults.html).
Additionally, [addCountsOfFeatureClasses](../reference/addCountsOfFeatureClasses.html) is used to count the number of BLAST hits grouped by user defined feature classes.
Summary files are then written to the `out_dir` by [writeSummaryFiles](../reference/writeSummaryFiles.html).
```{r, eval=FALSE}
classes <- c("tRNA","[Hh]sa","^U")
summary <- mergeResults(sigResults, countStats, blastResult, clustResult, map)
summary <- addCountsOfFeatureClasses(summary, classes)
writeSummaryFiles(summary, out_dir)

```
