<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Run DEUS</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Run DEUS</h1>



<div id="optional-preprocessing-of-fastq-files" class="section level3">
<h3>(Optional) preprocessing of FASTQ files</h3>
<p>Before applying DEUS we strongly recommend to trim your input FASTQ files. <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a> can be used to do adapter and quality trimming. The <code>--max_length</code> parameter should be set to (read length - 1) to remove non-small RNA sequences.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># remove adapters (automatically recognized by Trim Galore)</span>
<span class="kw">/software/trim_galore_v0.x.x/trim_galore</span> -q 0 --length 16 --max_length 49 <span class="ot">$in_file</span> -o <span class="ot">$out_dir</span>
<span class="co"># quality trimming (use dummy adapter to suppress adapter trimming)</span>
<span class="kw">/software/trim_galore_v0.x.x/trim_galore</span> -q 20 --length 1 -a CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC --stringency 50 <span class="ot">$in_file</span> -o <span class="ot">$out_dir</span></code></pre></div>
</div>
<div id="installing-deus" class="section level3">
<h3>Installing DEUS</h3>
<p>You can install DEUS from GitHub with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw">require</span>(<span class="st">&quot;devtools&quot;</span>)) <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools::<span class="kw">install_github</span>(<span class="st">&quot;timjeske/DEUS&quot;</span>)</code></pre></div>
<pre><code>## 
##   
   checking for file ‘/tmp/Rtmptx0RNm/remotes12df64a21f49/timjeske-DEUS-4520471/DESCRIPTION’ ...
  
✔  checking for file ‘/tmp/Rtmptx0RNm/remotes12df64a21f49/timjeske-DEUS-4520471/DESCRIPTION’
## 
  
─  preparing ‘DEUS’:
## 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
## 
  
─  checking for LF line-endings in source and make files and shell scripts
## 
  
─  checking for empty or unneeded directories
## 
  
─  building ‘DEUS_1.0.tar.gz’
## 
  
   Warning: invalid uid value replaced by that for user 'nobody'
##    Warning: invalid gid value replaced by that for user 'nobody'
## 
  
   
## </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DEUS)</code></pre></div>
<p>Alternatively, you can install DEUS from a local copy.</p>
<p>First, you need to unzip it:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">unzip</span> DEUS-master.zip</code></pre></div>
<p>Then you can install it in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw">require</span>(<span class="st">&quot;devtools&quot;</span>)) <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools::<span class="kw">install_local</span>(<span class="st">&quot;./DEUS-master&quot;</span>)
<span class="kw">library</span>(DEUS)</code></pre></div>
</div>
<div id="setting-parameters" class="section level3">
<h3>Setting parameters</h3>
<p>Before running the individual steps of the DEUS pipeline it is helpful to set some required parameters. The <code>in_dir</code> is the folder where the (trimmed) FASTQ files are stored. The <code>out_dir</code> is the folder where plots of differential expression analysis (DEA), clustering and summary output files will be stored. The <code>phenofile</code> is the condition file for DEA (<a href="DEUS.html#preparing-a-condition-file">examplary condition file</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">in_dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;DEUS&quot;</span>)
out_dir &lt;-<span class="st"> &quot;~/tmp&quot;</span>
phenofile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;condition.tsv&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;DEUS&quot;</span>)
pheno_info &lt;-<span class="st"> </span><span class="kw">read.table</span>(phenofile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">check.names=</span><span class="ot">FALSE</span>)
pheno_info</code></pre></div>
<pre><code>##                      sample condition
## cond1_s1 cond1_s1.fwd.fq.gz     cond1
## cond1_s2 cond1_s2.fwd.fq.gz     cond1
## cond1_s3 cond1_s3.fwd.fq.gz     cond1
## cond1_s4 cond1_s4.fwd.fq.gz     cond1
## cond1_s5 cond1_s5.fwd.fq.gz     cond1
## cond2_s1 cond2_s1.fwd.fq.gz     cond2
## cond2_s2 cond2_s2.fwd.fq.gz     cond2
## cond2_s3 cond2_s3.fwd.fq.gz     cond2
## cond2_s4 cond2_s4.fwd.fq.gz     cond2
## cond2_s5 cond2_s5.fwd.fq.gz     cond2</code></pre>
</div>
<div id="counting-unique-sequences" class="section level3">
<h3>Counting unique sequences</h3>
<p>To create the count table for differential expression analysis, <a href="../reference/createCountTableFromFastQs.html">createCountTableFromFastQs</a> counts all unique sequences in each FASTQ file given in your <code>pheno_info</code>. To remove low expressed sequences and reduce the size of the count table <a href="../reference/filterLowExp.html">filterLowExp</a> is applied. Sequences are kept if all groups of samples have a default minimum average sequence count of 10 or if one of the groups has zero counts for all samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count_table &lt;-<span class="st"> </span><span class="kw">createCountTableFromFastQs</span>(in_dir, <span class="dt">pheno_info=</span>pheno_info)
count_table_filtered &lt;-<span class="st"> </span><span class="kw">filterLowExp</span>(count_table, pheno_info)
<span class="kw">head</span>(count_table_filtered, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           cond1_s1 cond1_s2 cond1_s3 cond1_s4 cond1_s5
## CTCAGCCTTTTGGCTAAGATCAAGT       20        9       13       20       17
## CTTCTCAGCCTTTTGGCTAAGATCA       17       10       14        9       16
##                           cond2_s1 cond2_s2 cond2_s3 cond2_s4 cond2_s5
## CTCAGCCTTTTGGCTAAGATCAAGT        9       13        3       14       12
## CTTCTCAGCCTTTTGGCTAAGATCA       23       19       13       14       14</code></pre>
</div>
<div id="create-identifiers-for-each-sequence" class="section level3">
<h3>Create identifiers for each sequence</h3>
<p>A map is created by <a href="../reference/createMap.html">createMap</a> which assigns a unique identifier to each sequence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map &lt;-<span class="st"> </span><span class="kw">createMap</span>(count_table_filtered)
<span class="kw">head</span>(map, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID
## CTCAGCCTTTTGGCTAAGATCAAGT      seq_1
## CTTCTCAGCCTTTTGGCTAAGATCA      seq_2</code></pre>
</div>
<div id="running-differential-expression-analysis" class="section level3">
<h3>Running differential expression analysis</h3>
<p>Differential expression analysis is done by calling the function <a href="../reference/runDESeq2.html">runDESeq2</a>. Results are filtered using a 0.05 Pvalue cutoff.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>condition
de_results &lt;-<span class="st"> </span><span class="kw">runDESeq2</span>(count_table_filtered, pheno_info, design, <span class="dt">out_dir=</span>out_dir)
sig_results &lt;-<span class="st"> </span>de_results$de_result
sig_results &lt;-<span class="st"> </span>sig_results[!<span class="kw">is.na</span>(sig_results$IHWPvalue) &amp;<span class="st"> </span>sig_results$IHWPvalue &lt;<span class="st"> </span><span class="fl">0.05</span>,]
<span class="kw">head</span>(sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                                Pvalue       padj baseMean Log2FoldChange
## CTCAGCCTTTTGGCTAAGATCAAGT 0.020355991 0.02424650 12.18654     -0.7844677
## TAGTAGGGGACTGCGTTCCCACTTT 0.001960278 0.01353789  1.51890     -3.5383764
##                               lfcSE      stat  IHWPvalue
## CTCAGCCTTTTGGCTAAGATCAAGT 0.3381733 -2.319721 0.02424650
## TAGTAGGGGACTGCGTTCCCACTTT 1.1428180 -3.096185 0.01353789</code></pre>
</div>
<div id="running-blast" class="section level3">
<h3>Running BLAST</h3>
<p>The significant sequences are subsequently stored in a FASTA file for BLAST annotation using <a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a>. BLAST is executed via <a href="../reference/runBlast.html">runBlast</a>. It is necessary to set the <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST binary</a>, your <a href="DEUS.html#preparing-a-blast-database">BLAST database</a> and the number of threads to be used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_exec &lt;-<span class="st"> &quot;/your/path/to/ncbi-blast-x.x.x/bin/blastn&quot;</span>
blast_db &lt;-<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;blastdb/DASHR_subset.fa&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;DEUS&quot;</span>)
blast_ncores &lt;-<span class="st"> </span><span class="dv">2</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw">sequencesAsFasta</span>(sig_results,map)
blast_result &lt;-<span class="st"> </span><span class="kw">runBlast</span>(blast_exec, blast_db, blast_ncores, sig_seq_fasta) </code></pre></div>
<p>In our example, we load pre-compiled data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_int &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;results/blast_result_sig_sequences.tsv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;DEUS&quot;</span>)
blast_result &lt;-<span class="st"> </span><span class="kw">read.table</span>(blast_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">head</span>(blast_result)</code></pre></div>
<pre><code>##   SequenceID Annotation Length BlastEvalue
## 1      seq_1   U2-L1052     25    8.23e-09
## 2      seq_1   U2-L1045     25    8.23e-09
## 3      seq_1   U2-L1006     25    8.23e-09
## 4      seq_1    U2-L996     25    8.23e-09
## 5      seq_1    U2-L976     25    8.23e-09
## 6      seq_1    U2-L970     25    8.23e-09</code></pre>
</div>
<div id="compute-counts-per-condition" class="section level3">
<h3>Compute counts per condition</h3>
<p>The results of <a href="../reference/runDESeq2.html">runDESeq2</a> are used as input for <a href="../reference/getConditionCountStats.html">getConditionCountStats</a> to compute the mean and the standard deviation of the normalized counts for each analyzed condition.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count_stats &lt;-<span class="st"> </span><span class="kw">getConditionCountStats</span>(de_results$norm_counts, pheno_info)
<span class="kw">head</span>(count_stats, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           NormCounts_cond1_Mean NormCounts_cond1_Sd
## CTCAGCCTTTTGGCTAAGATCAAGT              15.45875            5.266979
## CTTCTCAGCCTTTTGGCTAAGATCA              12.72590            2.984202
##                           NormCounts_cond2_Mean NormCounts_cond2_Sd
## CTCAGCCTTTTGGCTAAGATCAAGT              8.914326            4.224273
## CTTCTCAGCCTTTTGGCTAAGATCA             14.217300            1.156204</code></pre>
</div>
<div id="run-clustering" class="section level3">
<h3>Run clustering</h3>
<p>The CD-Hit algorithm is executed via <a href="../reference/runClustering.html">runClustering</a>. The function requires the <a href="http://weizhongli-lab.org/cd-hit/">cd-hit-est</a> binary.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cd_hit &lt;-<span class="st"> &quot;/your/path/to/cdhit/cd-hit-est&quot;</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw">sequencesAsFasta</span>(sig_results,map)
clust_result &lt;-<span class="st"> </span><span class="kw">runClustering</span>(cd_hit, sig_seq_fasta, out_dir, <span class="dt">identity_cutoff=</span><span class="fl">0.9</span>, <span class="dt">length_cutoff=</span><span class="fl">0.9</span>, <span class="dt">wordlength=</span><span class="dv">9</span>, map)</code></pre></div>
<p>For this manual, we will load available example data again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clust_int &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;results/clust_result_sig_sequences.tsv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;DEUS&quot;</span>)
clust_result &lt;-<span class="st"> </span><span class="kw">read.table</span>(clust_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">rownames</span>(clust_result) &lt;-<span class="st"> </span>clust_result$SequenceID
<span class="kw">head</span>(clust_result)</code></pre></div>
<pre><code>##          SequenceID ClusterID
## seq_1         seq_1         0
## seq_1003   seq_1003         1
## seq_1010   seq_1010       203
## seq_1011   seq_1011        27
## seq_102     seq_102        37
## seq_1026   seq_1026       204</code></pre>
</div>
<div id="merging-the-results" class="section level3">
<h3>Merging the results</h3>
<p>Finally, the group-wise mean and standard deviation of the normalized counts, the results of DEA, BLASTing and clustering are merged together by <a href="../reference/mergeResults.html">mergeResults</a>. Additionally, <a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a> is used to count the number of BLAST hits grouped by user defined feature classes. Summary files are then written to the <code>out_dir</code> by <a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tRNA&quot;</span>,<span class="st">&quot;[Hh]sa&quot;</span>,<span class="st">&quot;^U&quot;</span>)
summary &lt;-<span class="st"> </span><span class="kw">mergeResults</span>(sig_results, count_stats, blast_result, clust_result, map)
summary &lt;-<span class="st"> </span><span class="kw">addCountsOfFeatureClasses</span>(summary, classes)
<span class="kw">writeSummaryFiles</span>(summary, out_dir)
<span class="kw">head</span>(summary, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID Log2FoldChange       Pvalue
## CACCCTGATCACACTCCATCTAGTC    seq_238      -4.011708 0.0001079705
## TTGACTGATCTTGGAAGCTAAGCAG    seq_458      -3.987188 0.0001261311
##                            IHWPvalue NormCounts_cond1_Mean
## CACCCTGATCACACTCCATCTAGTC 0.01353789              3.938940
## TTGACTGATCTTGGAAGCTAAGCAG 0.01353789              3.860569
##                           NormCounts_cond1_Sd NormCounts_cond2_Mean
## CACCCTGATCACACTCCATCTAGTC            1.263952                     0
## TTGACTGATCTTGGAAGCTAAGCAG            1.366233                     0
##                           NormCounts_cond2_Sd ClusterID Length BlastEvalue
## CACCCTGATCACACTCCATCTAGTC                   0        84     25          NA
## TTGACTGATCTTGGAAGCTAAGCAG                   0       103     25          NA
##                           tRNA [Hh]sa ^U Other FeatureList
## CACCCTGATCACACTCCATCTAGTC    0      0  0     0          NA
## TTGACTGATCTTGGAAGCTAAGCAG    0      0  0     0          NA</code></pre>
</div>
<div id="computing-the-na-fraction" class="section level3">
<h3>Computing the NA fraction</h3>
<p>The function <a href="../reference/getNoBlastHitFraction.html">getNoBlastHitFraction</a> can be used to compute the fraction of reads without annotation for each sample. This fraction corresponds to the fraction of reads that are likely to be ignored in mapping approaches because they map to regions that are not annotated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getNoBlastHitFraction</span>(summary, de_results$norm_counts)</code></pre></div>
<pre><code>##          NA_fraction
## cond1_s1  0.14445097
## cond1_s2  0.13165769
## cond1_s3  0.15236318
## cond1_s4  0.14094351
## cond1_s5  0.13832335
## cond2_s1  0.05046154
## cond2_s2  0.04441703
## cond2_s3  0.05305367
## cond2_s4  0.05405405
## cond2_s5  0.06030484</code></pre>
<hr />
</div>
<div id="extended-expression-analysis-using-sequence-clusters" class="section level3">
<h3>Extended expression analysis using sequence clusters</h3>
<p>In addition to the default DEUS pipeline, we developed an adjusted approach that aggregates sequence counts for each sequence cluster and uses this information to find differentially expressed sequence clusters. By some small adjustments, the cluster expression analysis can be included into the default DEUS pipeline.</p>
<p>Instead of creating the sequence map for differentially expressed sequences only, it will be created on the raw input sequences since all sequences will be used for clustering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map &lt;-<span class="st"> </span><span class="kw">createMap</span>(count_table)</code></pre></div>
<p>Afterwards, all sequences can be clustered. Based on the clustering results, the sum of sequence counts will be created for each cluster via <a href="../reference/mergeAndAggregate.html">mergeAndAggregate</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cd_hit &lt;-<span class="st"> &quot;/your/path/to/cdhit/cd-hit-est&quot;</span>
all_seq_fasta &lt;-<span class="st"> </span><span class="kw">sequencesAsFasta</span>(count_table,map)
clust_result &lt;-<span class="st"> </span><span class="kw">runClustering</span>(cd_hit, all_seq_fasta, out_dir, <span class="dt">identity_cutoff=</span><span class="fl">0.9</span>, <span class="dt">length_cutoff=</span><span class="fl">0.9</span>, <span class="dt">wordlength=</span><span class="dv">9</span>, map)
cl_counts &lt;-<span class="st"> </span><span class="kw">mergeAndAggregate</span>(map, count_table, clust_result)</code></pre></div>
<p>For this manual, we will load the provided clustering results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clust_int &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;results/clust_result_clust_sequences.tsv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;DEUS&quot;</span>)
clust_result &lt;-<span class="st"> </span><span class="kw">read.table</span>(clust_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">rownames</span>(clust_result) &lt;-<span class="st"> </span>clust_result$SequenceID
cl_counts &lt;-<span class="st"> </span><span class="kw">mergeAndAggregate</span>(map, count_table, clust_result)
<span class="kw">head</span>(cl_counts, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##   cond1_s1 cond1_s2 cond1_s3 cond1_s4 cond1_s5 cond2_s1 cond2_s2 cond2_s3
## 0      111       63       94       90       95      128      110       96
## 1       76       77       77       79       56       89       93       96
##   cond2_s4 cond2_s5
## 0       90      100
## 1       79       80</code></pre>
<p>The resulting count matrix is then used as input for the DE analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>condition
cl_de_results &lt;-<span class="st"> </span><span class="kw">runDESeq2</span>(cl_counts, pheno_info, design, <span class="dt">out_dir =</span> out_dir, <span class="dt">prefix =</span> <span class="st">&quot;Cluster&quot;</span>)
cl_sig_results &lt;-<span class="st"> </span>cl_de_results$de_result
<span class="kw">head</span>(cl_sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##       Pvalue      padj baseMean Log2FoldChange     lfcSE     stat
## 0 0.18160997 0.3911030 92.35134      0.1918463 0.1436175 1.335814
## 1 0.09000106 0.2178087 76.00927      0.2394114 0.1412130 1.695392
##    IHWPvalue
## 0 0.15022735
## 1 0.08029553</code></pre>
<p>Subsequently, the existing DEA results based on individual sequence counts and the generated DEA results based on sequence clusters can be merged into a single data frame via <a href="../reference/mergeSingleAndClusterResults.html">mergeSingleAndClusterResults</a>. Based on the individual and cluster Pvalues, the resulting data frame is filtered to detect sequences that are differentially expressed based on their sequence count or part of a differentially expressed sequence cluster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sig_results &lt;-<span class="st"> </span><span class="kw">mergeSingleAndClusterResults</span>(cl_sig_results,clust_result,sig_results,map)
sig_results &lt;-<span class="st"> </span>sig_results[((!<span class="kw">is.na</span>(sig_results$IHWPval) &amp;<span class="st"> </span>sig_results$IHWPval &lt;<span class="st"> </span><span class="fl">0.05</span>) |<span class="st"> </span>(!<span class="kw">is.na</span>(sig_results$Cl_IHWPval) &amp;<span class="st"> </span>sig_results$Cl_IHWPval &lt;<span class="st"> </span><span class="fl">0.05</span>)),]
<span class="kw">head</span>(sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                                Pvalue       padj baseMean Log2FoldChange
## CTCAGCCTTTTGGCTAAGATCAAGT 0.020355991 0.02424650 12.18654     -0.7844677
## TAGTAGGGGACTGCGTTCCCACTTT 0.001960278 0.01353789  1.51890     -3.5383764
##                               lfcSE      stat  IHWPvalue SequenceID
## CTCAGCCTTTTGGCTAAGATCAAGT 0.3381733 -2.319721 0.02424650      seq_1
## TAGTAGGGGACTGCGTTCCCACTTT 1.1428180 -3.096185 0.01353789     seq_63
##                              Cl_Pvalue      Cl_padj Cl_baseMean
## CTCAGCCTTTTGGCTAAGATCAAGT 1.816100e-01 3.911030e-01   92.351345
## TAGTAGGGGACTGCGTTCCCACTTT 2.441897e-07 2.315223e-05    5.724111
##                           Cl_Log2FoldChange  Cl_lfcSE   Cl_stat
## CTCAGCCTTTTGGCTAAGATCAAGT         0.1918463 0.1436175  1.335814
## TAGTAGGGGACTGCGTTCCCACTTT        -5.5011357 1.0656770 -5.162104
##                           Cl_IHWPvalue ClusterID
## CTCAGCCTTTTGGCTAAGATCAAGT 1.502273e-01         0
## TAGTAGGGGACTGCGTTCCCACTTT 1.840328e-05        38</code></pre>
<p>Next, we run BLAST on all selected sequences.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_exec &lt;-<span class="st"> &quot;/your/path/to/ncbi-blast-x.x.x/bin/blastn&quot;</span>
blast_db &lt;-<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;blastdb/DASHR_subset.fa&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;DEUS&quot;</span>)
blast_ncores &lt;-<span class="st"> </span><span class="dv">2</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw">sequencesAsFasta</span>(sig_results,map)
blast_result &lt;-<span class="st"> </span><span class="kw">runBlast</span>(blast_exec, blast_db, blast_ncores, sig_seq_fasta) </code></pre></div>
<p>In our example, we load pre-compiled data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_int &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;results/blast_result_clust_sequences.tsv&quot;</span>, <span class="dt">package=</span><span class="st">&quot;DEUS&quot;</span>)
blast_result &lt;-<span class="st"> </span><span class="kw">read.table</span>(blast_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">head</span>(blast_result)</code></pre></div>
<pre><code>##   SequenceID Annotation Length BlastEvalue
## 1      seq_1   U2-L1052     25    8.23e-09
## 2      seq_1   U2-L1045     25    8.23e-09
## 3      seq_1   U2-L1006     25    8.23e-09
## 4      seq_1    U2-L996     25    8.23e-09
## 5      seq_1    U2-L976     25    8.23e-09
## 6      seq_1    U2-L970     25    8.23e-09</code></pre>
<p>Finally, the summary table is generated again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># merge results</span>
classes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tRNA&quot;</span>,<span class="st">&quot;[Hh]sa&quot;</span>,<span class="st">&quot;^U&quot;</span>)
summary &lt;-<span class="st"> </span><span class="kw">mergeResults</span>(sig_results, count_stats, blast_result, clust_result, map)
summary &lt;-<span class="st"> </span><span class="kw">addCountsOfFeatureClasses</span>(summary, classes)
<span class="kw">writeSummaryFiles</span>(summary, out_dir)
<span class="kw">head</span>(summary, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID Log2FoldChange       Pvalue
## CACCCTGATCACACTCCATCTAGTC   seq_1105      -4.011708 0.0001079705
## TTGACTGATCTTGGAAGCTAAGCAG   seq_1971      -3.987188 0.0001261311
##                            IHWPvalue Cl_Log2FoldChange    Cl_Pvalue
## CACCCTGATCACACTCCATCTAGTC 0.01353789         -5.078722 2.487141e-07
## TTGACTGATCTTGGAAGCTAAGCAG 0.01353789         -1.701923 2.261974e-05
##                           Cl_IHWPvalue NormCounts_cond1_Mean
## CACCCTGATCACACTCCATCTAGTC 5.232667e-05              3.938940
## TTGACTGATCTTGGAAGCTAAGCAG 2.970498e-04              3.860569
##                           NormCounts_cond1_Sd NormCounts_cond2_Mean
## CACCCTGATCACACTCCATCTAGTC            1.263952                     0
## TTGACTGATCTTGGAAGCTAAGCAG            1.366233                     0
##                           NormCounts_cond2_Sd ClusterID Length BlastEvalue
## CACCCTGATCACACTCCATCTAGTC                   0       820     25          NA
## TTGACTGATCTTGGAAGCTAAGCAG                   0      1050     25          NA
##                           tRNA [Hh]sa ^U Other FeatureList
## CACCCTGATCACACTCCATCTAGTC    0      0  0     0          NA
## TTGACTGATCTTGGAAGCTAAGCAG    0      0  0     0          NA</code></pre>
<p>In order to compare both aproaches, the function <a href="../reference/printClusterSummary.html">printClusterSummary</a> can be used to get detailed information about the number of significant sequences and clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">printClusterSummary</span>(summary)</code></pre></div>
<pre><code>## ## Sequence summary ##
## Significant sequences: 1119
## Significant sequences not clustered: 0
## Sequences in significant clusters: 18508
## Non-significant sequences in significant clusters: 17523
## - too low expressed: 17523
## - too weak differentially expressed: 0
## Significant sequences in non-significant clusters: 134
## ## Cluster summary ##
## Significant clusters: 3675
## Clusters including significant sequences: 821
## Significant clusters including no significant sequences: 2959</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
