<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Run DEUS â€¢ DEUS</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Run DEUS">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DEUS</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/DEUS.html">Get started</a>
</li>
<li>
  <a href="../articles/SampleUsage.html">Run DEUS</a>
</li>
<li>
  <a href="../reference/index.html">R functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/timjeske/DEUS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Run DEUS</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/timjeske/DEUS/blob/master/vignettes/SampleUsage.Rmd"><code>vignettes/SampleUsage.Rmd</code></a></small>
      <div class="hidden name"><code>SampleUsage.Rmd</code></div>

    </div>

    
    
<div id="optional-preprocessing-of-fastq-files" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-preprocessing-of-fastq-files" class="anchor"></a>(Optional) preprocessing of FASTQ files</h3>
<p>Before applying DEUS we strongly recommend to trim your input FASTQ files. <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a> can be used to do adapter and quality trimming. The <code>--max_length</code> parameter should be set to (read length - 1) to remove non-small RNA sequences.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># remove adapters (automatically recognized by Trim Galore)</span>
<span class="kw">/software/trim_galore_v0.x.x/trim_galore</span> -q 0 --length 16 --max_length 49 <span class="ot">$in_file</span> -o <span class="ot">$out_dir</span>
<span class="co"># quality trimming (use dummy adapter to suppress adapter trimming)</span>
<span class="kw">/software/trim_galore_v0.x.x/trim_galore</span> -q 20 -a CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC --stringency 50 <span class="ot">$in_file</span> -o <span class="ot">$out_dir</span></code></pre></div>
</div>
<div id="installing-deus" class="section level3">
<h3 class="hasAnchor">
<a href="#installing-deus" class="anchor"></a>Installing DEUS</h3>
<p>You can install DEUS from GitHub with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw">require</span>(<span class="st">"devtools"</span>)) <span class="kw">install.packages</span>(<span class="st">"devtools"</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools::<span class="kw">install_github</span>(<span class="st">"timjeske/DEUS"</span>)
<span class="kw">library</span>(DEUS)</code></pre></div>
<p>Alternatively, you can install DEUS from a local copy.</p>
<p>First, you need to unzip it:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">unzip</span> DEUS-master.zip</code></pre></div>
<p>Then you can install it in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw">require</span>(<span class="st">"devtools"</span>)) <span class="kw">install.packages</span>(<span class="st">"devtools"</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools::<span class="kw">install_local</span>(<span class="st">"./DEUS-master"</span>)
<span class="kw">library</span>(DEUS)</code></pre></div>
</div>
<div id="setting-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-parameters" class="anchor"></a>Setting parameters</h3>
<p>Before running the individual steps of the DEUS pipeline it is helpful to set all required parameters. The <code>in_dir</code> is the folder where the (trimmed) FASTQ files are stored. The <code>phenofile</code> is the condition file for differential expression analysis (DEA) (<a href="DEUS.html#preparing-a-condition-file">examplary condition file</a>). The <code>out_dir</code> is the folder where DEA plots, clustering and summary output files will be stored. To use BLAST, it is necessary to set the <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST binary</a>, your <a href="DEUS.html#preparing-a-blast-database">BLAST database</a> and the number of threads to be used. To cluster resulting sequencing, the <a href="http://weizhongli-lab.org/cd-hit/">cd-hit-est</a> binary is reuqired.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">in_dir &lt;-<span class="st"> "/data/fastq-files/"</span>
phenofile &lt;-<span class="st"> "/data/condition.tsv"</span>

out_dir &lt;-<span class="st"> "data/results"</span>

blast_exec &lt;-<span class="st"> "/software/blast/bin/blastn"</span>
blast_db &lt;-<span class="st"> "/data/My_Blast_DB.fa"</span>
blast_ncores &lt;-<span class="st"> </span><span class="dv">2</span>

cd_hit &lt;-<span class="st"> "/software/cdhit/cd-hit-est"</span></code></pre></div>
</div>
<div id="counting-unique-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-unique-sequences" class="anchor"></a>Counting unique sequences</h3>
<p>To create the count table for differential expression analysis, <a href="../reference/createCountTableFromFastQs.html">createCountTableFromFastQs</a> counts all unique sequences in each FASTQ file given in your <code>phenofile</code>. To remove low expressed sequences and reduce the size of the count table <a href="../reference/filterLowExp.html">filterLowExp</a> is applied. Sequences are kept if both groups of samples have a default minimum average sequence count of 10 or if one of both groups has zero counts for all samples. It can be useful to store the filtered count table in our output folder as counting takes some time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">phenoInfo &lt;-<span class="st"> </span><span class="kw">read.table</span>(phenofile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">check.names=</span><span class="ot">FALSE</span>)
countTable &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createCountTableFromFastQs.html">createCountTableFromFastQs</a></span>(in_dir, <span class="dt">phenoInfo=</span>phenoInfo)
countTable &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterLowExp.html">filterLowExp</a></span>(countTable, phenoInfo)
<span class="kw">write.table</span>(countTable, <span class="kw">paste</span>(out_dir,<span class="st">"AllCounts_filtered.tsv"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">col.names=</span>T, <span class="dt">quote=</span>F, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">row.names=</span>T)</code></pre></div>
</div>
<div id="running-differential-expression-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#running-differential-expression-analysis" class="anchor"></a>Running differential expression analysis</h3>
<p>Differential expression analysis is done by calling the function <a href="../reference/runDESeq2.html">runDESeq2</a>. After filtering for significant sequences, a map is created by <a href="../reference/createMap.html">createMap</a> which assigns a unique identifier to each sequence. The results of <a href="../reference/runDESeq2.html">runDESeq2</a> are additionally used as input for <a href="../reference/getConditionCountStats.html">getConditionCountStats</a> to compute the mean and the standard deviation of the normalized counts for each analyzed conditions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>condition
deResults &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runDESeq2.html">runDESeq2</a></span>(countTable, phenoInfo, design, map, out_dir)
sigResults &lt;-<span class="st"> </span>deResults$deResult
sigResults &lt;-<span class="st"> </span>sigResults[!<span class="kw">is.na</span>(sigResults$IHWPval) &amp;<span class="st"> </span>sigResults$IHWPval &lt;<span class="st"> </span><span class="fl">0.05</span>,]
map &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createMap.html">createMap</a></span>(sigResults)

<span class="co"># get count stats</span>
countStats &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getConditionCountStats.html">getConditionCountStats</a></span>(deResults$normCounts, phenoInfo)</code></pre></div>
</div>
<div id="running-blast" class="section level3">
<h3 class="hasAnchor">
<a href="#running-blast" class="anchor"></a>Running BLAST</h3>
<p>To annotate significant sequences they need to transformed to a vector of sequences in FASTA style by using <a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a>. Afterwards BLAST is executed by <a href="../reference/runBlast.html">runBlast</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sigSeqFasta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a></span>(sigResults,map)
blastResult &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runBlast.html">runBlast</a></span>(blast_exec, blast_db, blast_ncores, sigSeqFasta)</code></pre></div>
</div>
<div id="run-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#run-clustering" class="anchor"></a>Run clustering</h3>
<p>The CD-Hit algorithm is executed via <a href="../reference/runClustering.html">runClustering</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  clustResult &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runClustering.html">runClustering</a></span>(cd_hit, sigSeqFasta, out_dir, <span class="fl">0.9</span>, <span class="fl">0.9</span>, <span class="dv">9</span>, map)</code></pre></div>
</div>
<div id="merging-the-results" class="section level3">
<h3 class="hasAnchor">
<a href="#merging-the-results" class="anchor"></a>Merging the results</h3>
<p>Finally, the group-wise mean and standard deviation of the normalized counts, the results of DEA, BLASTing and clustering are merged together by <a href="../reference/mergeResults.html">mergeResults</a>. Additionally, <a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a> is used to count the number of BLAST hits grouped by user defined feature classes. Summary files are then written to the <code>out_dir</code> by <a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"tRNA"</span>,<span class="st">"[Hh]sa"</span>,<span class="st">"^U"</span>)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeResults.html">mergeResults</a></span>(sigResults, countStats, blastResult, clustResult, map)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a></span>(summary, classes)
<span class="kw"><a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a></span>(summary, out_dir)</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Tim Jeske, Maximilian Hastreiter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
