<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Run DEUS â€¢ DEUS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Run DEUS">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DEUS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/DEUS.html">Get started</a>
</li>
<li>
  <a href="../articles/SampleUsage.html">Run DEUS</a>
</li>
<li>
  <a href="../reference/index.html">R functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/timjeske/DEUS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Run DEUS</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/timjeske/DEUS/blob/master/vignettes/SampleUsage.Rmd"><code>vignettes/SampleUsage.Rmd</code></a></small>
      <div class="hidden name"><code>SampleUsage.Rmd</code></div>

    </div>

    
    
<div id="optional-preprocessing-of-fastq-files" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-preprocessing-of-fastq-files" class="anchor"></a>(Optional) preprocessing of FASTQ files</h3>
<p>Before applying DEUS we strongly recommend to trim your input FASTQ files. <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a> can be used to do adapter and quality trimming. The <code>--max_length</code> parameter should be set to (read length - 1) to remove non-small RNA sequences.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># remove adapters (automatically recognized by Trim Galore)</span>
<span class="ex">/software/trim_galore_v0.x.x/trim_galore</span> -q 0 --length 16 --max_length 49 <span class="va">$in_file</span> -o <span class="va">$out_dir</span>
<span class="co"># quality trimming (use dummy adapter to suppress adapter trimming)</span>
<span class="ex">/software/trim_galore_v0.x.x/trim_galore</span> -q 20 --length 1 -a CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC --stringency 50 <span class="va">$in_file</span> -o <span class="va">$out_dir</span></code></pre></div>
</div>
<div id="installing-deus" class="section level3">
<h3 class="hasAnchor">
<a href="#installing-deus" class="anchor"></a>Installing DEUS</h3>
<p>You can install DEUS from GitHub with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"devtools"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"timjeske/DEUS"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(DEUS)</code></pre></div>
<p>Alternatively, you can install DEUS from a local copy.</p>
<p>First, you need to unzip it:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">unzip</span> DEUS-master.zip</code></pre></div>
<p>Then you can install it in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"devtools"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>, <span class="dt">repos=</span><span class="st">'http://cran.us.r-project.org'</span>)
devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_local</a></span>(<span class="st">"./DEUS-master"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(DEUS)</code></pre></div>
</div>
<div id="setting-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-parameters" class="anchor"></a>Setting parameters</h3>
<p>Before running the individual steps of the DEUS pipeline it is helpful to set some required parameters. The <code>in_dir</code> is the folder where the (trimmed) FASTQ files are stored. The <code>out_dir</code> is the folder where plots of differential expression analysis (DEA), clustering and summary output files will be stored. The <code>phenofile</code> is the condition file for DEA (<a href="DEUS.html#preparing-a-condition-file">examplary condition file</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">in_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"DEUS"</span>)
out_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempdir</a></span>()
phenofile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"condition.tsv"</span>, <span class="dt">package =</span> <span class="st">"DEUS"</span>)
pheno_info &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(phenofile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">check.names=</span><span class="ot">FALSE</span>)
pheno_info</code></pre></div>
<pre><code>##                      sample condition
## cond1_s1 cond1_s1.fwd.fq.gz     cond1
## cond1_s2 cond1_s2.fwd.fq.gz     cond1
## cond1_s3 cond1_s3.fwd.fq.gz     cond1
## cond1_s4 cond1_s4.fwd.fq.gz     cond1
## cond1_s5 cond1_s5.fwd.fq.gz     cond1
## cond2_s1 cond2_s1.fwd.fq.gz     cond2
## cond2_s2 cond2_s2.fwd.fq.gz     cond2
## cond2_s3 cond2_s3.fwd.fq.gz     cond2
## cond2_s4 cond2_s4.fwd.fq.gz     cond2
## cond2_s5 cond2_s5.fwd.fq.gz     cond2</code></pre>
</div>
<div id="counting-unique-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-unique-sequences" class="anchor"></a>Counting unique sequences</h3>
<p>To create the count table for differential expression analysis, <a href="../reference/createCountTableFromFastQs.html">createCountTableFromFastQs</a> counts all unique sequences in each FASTQ file given in your <code>pheno_info</code>. As an optional step, we only keep sequences with an average sequence count larger than 1 to remove low expressed sequences and reduce the size of the count table.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count_table &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createCountTableFromFastQs.html">createCountTableFromFastQs</a></span>(in_dir, <span class="dt">pheno_info=</span>pheno_info)
count_table_filtered &lt;-<span class="st"> </span>count_table[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowMeans</a></span>(count_table)<span class="op">&gt;</span><span class="dv">1</span>,]
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(count_table_filtered, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           cond1_s1 cond1_s2 cond1_s3 cond1_s4 cond1_s5
## TCTCAGCCTTTTGGCTAAGATCAAG       19       19       21       10       18
## CTCAGCCTTTTGGCTAAGATCAAGT       16       19       13       21       17
##                           cond2_s1 cond2_s2 cond2_s3 cond2_s4 cond2_s5
## TCTCAGCCTTTTGGCTAAGATCAAG       13       10       15       22       10
## CTCAGCCTTTTGGCTAAGATCAAGT       12        6       16       19       10</code></pre>
</div>
<div id="create-identifiers-for-each-sequence" class="section level3">
<h3 class="hasAnchor">
<a href="#create-identifiers-for-each-sequence" class="anchor"></a>Create identifiers for each sequence</h3>
<p>A map is created by <a href="../reference/createMap.html">createMap</a> which assigns a unique identifier to each sequence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createMap.html">createMap</a></span>(count_table_filtered)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(map, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID
## TCTCAGCCTTTTGGCTAAGATCAAG      seq_1
## CTCAGCCTTTTGGCTAAGATCAAGT      seq_2</code></pre>
</div>
<div id="running-differential-expression-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#running-differential-expression-analysis" class="anchor"></a>Running differential expression analysis</h3>
<p>Differential expression analysis is done by calling the function <a href="../reference/runDESeq2.html">runDESeq2</a>. Results are filtered using a 0.05 Pvalue cutoff.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>condition
de_results &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runDESeq2.html">runDESeq2</a></span>(count_table_filtered, pheno_info, design, <span class="dt">out_dir=</span>out_dir)
sig_results &lt;-<span class="st"> </span>de_results<span class="op">$</span>de_result
sig_results &lt;-<span class="st"> </span>sig_results[<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(sig_results<span class="op">$</span>IHWPvalue) <span class="op">&amp;</span><span class="st"> </span>sig_results<span class="op">$</span>IHWPvalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>,]
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                                 Pvalue       padj baseMean Log2FoldChange
## TCGCTTCTCAGCCTTTTGGCTAAGA 0.0007176335 0.03462413 7.263225      -1.521815
## CGCTTCTCAGCCTTTTGGCTAAGAT 0.0026808638 0.05501307 6.594336      -1.431923
##                               lfcSE      stat  IHWPvalue
## TCGCTTCTCAGCCTTTTGGCTAAGA 0.4498748 -3.382752 0.04498802
## CGCTTCTCAGCCTTTTGGCTAAGAT 0.4769671 -3.002143 0.04498802</code></pre>
</div>
<div id="running-blast" class="section level3">
<h3 class="hasAnchor">
<a href="#running-blast" class="anchor"></a>Running BLAST</h3>
<p>The significant sequences are subsequently stored in a FASTA file for BLAST annotation using <a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a>. BLAST is executed via <a href="../reference/runBlast.html">runBlast</a>. It is necessary to set the <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST binary</a>, your <a href="DEUS.html#preparing-a-blast-database">BLAST database</a> and the number of threads to be used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_exec &lt;-<span class="st"> "/your/path/to/ncbi-blast-x.x.x/bin/blastn"</span>
blast_db &lt;-<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"blastdb/DASHR_subset.fa"</span>, <span class="dt">package =</span> <span class="st">"DEUS"</span>)
blast_ncores &lt;-<span class="st"> </span><span class="dv">2</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a></span>(sig_results,map)
blast_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runBlast.html">runBlast</a></span>(blast_exec, blast_db, blast_ncores, sig_seq_fasta) </code></pre></div>
<p>In our example, we load pre-compiled data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_int &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"results/blast_result_sig_sequences.tsv"</span>, <span class="dt">package=</span><span class="st">"DEUS"</span>)
blast_result &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(blast_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(blast_result)</code></pre></div>
<pre><code>##   SequenceID Annotation Length BlastEvalue
## 1      seq_6   U2-L1036     25    8.23e-09
## 2      seq_6    U2-L996     25    8.23e-09
## 3      seq_6    U2-L968     25    8.23e-09
## 4      seq_6    U2-L946     25    8.23e-09
## 5      seq_6    U2-L930     25    8.23e-09
## 6      seq_6    U2-L929     25    8.23e-09</code></pre>
</div>
<div id="compute-counts-per-condition" class="section level3">
<h3 class="hasAnchor">
<a href="#compute-counts-per-condition" class="anchor"></a>Compute counts per condition</h3>
<p>The results of <a href="../reference/runDESeq2.html">runDESeq2</a> are used as input for <a href="../reference/getConditionCountStats.html">getConditionCountStats</a> to compute the mean and the standard deviation of the normalized counts for each analyzed condition.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count_stats &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getConditionCountStats.html">getConditionCountStats</a></span>(de_results<span class="op">$</span>norm_counts, pheno_info)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(count_stats, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           NormCounts_cond1_Mean NormCounts_cond1_Sd
## TCTCAGCCTTTTGGCTAAGATCAAG              16.86876            4.147471
## CTCAGCCTTTTGGCTAAGATCAAGT              16.67663            2.914544
##                           NormCounts_cond2_Mean NormCounts_cond2_Sd
## TCTCAGCCTTTTGGCTAAGATCAAG              13.54612            4.701311
## CTCAGCCTTTTGGCTAAGATCAAGT              12.20912            4.944475</code></pre>
</div>
<div id="run-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#run-clustering" class="anchor"></a>Run clustering</h3>
<p>The CD-Hit algorithm is executed via <a href="../reference/runClustering.html">runClustering</a>. The function requires the <a href="http://weizhongli-lab.org/cd-hit/">cd-hit-est</a> binary.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cd_hit &lt;-<span class="st"> "/your/path/to/cdhit/cd-hit-est"</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a></span>(sig_results,map)
clust_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runClustering.html">runClustering</a></span>(cd_hit, sig_seq_fasta, out_dir, <span class="dt">identity_cutoff=</span><span class="fl">0.9</span>, <span class="dt">length_cutoff=</span><span class="fl">0.9</span>, <span class="dt">wordlength=</span><span class="dv">9</span>, map)</code></pre></div>
<p>For this manual, we will load available example data again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clust_int &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"results/clust_result_sig_sequences.tsv"</span>, <span class="dt">package=</span><span class="st">"DEUS"</span>)
clust_result &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(clust_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(clust_result) &lt;-<span class="st"> </span>clust_result<span class="op">$</span>SequenceID
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(clust_result)</code></pre></div>
<pre><code>##           SequenceID ClusterID
## seq_1000    seq_1000       170
## seq_10011  seq_10011       632
## seq_1002    seq_1002       171
## seq_10023  seq_10023       633
## seq_10034  seq_10034       564
## seq_10036  seq_10036       634</code></pre>
</div>
<div id="merging-the-results" class="section level3">
<h3 class="hasAnchor">
<a href="#merging-the-results" class="anchor"></a>Merging the results</h3>
<p>Finally, the group-wise mean and standard deviation of the normalized counts, the results of DEA, BLASTing and clustering are merged together by <a href="../reference/mergeResults.html">mergeResults</a>. Additionally, <a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a> is used to count the number of BLAST hits grouped by user defined feature classes. Summary files are then written to the <code>out_dir</code> by <a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"tRNA"</span>,<span class="st">"[Hh]sa"</span>,<span class="st">"^U"</span>)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeResults.html">mergeResults</a></span>(sig_results, count_stats, blast_result, clust_result, map)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a></span>(summary, classes)
<span class="kw"><a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a></span>(summary, out_dir)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(summary, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID Log2FoldChange       Pvalue
## TGCTCACTTTGGCAGCACATATACT  seq_12936       4.575022 3.738245e-07
## CTTCTCAGTCTTTTGGCTAAGATCA  seq_13006       4.446684 9.264864e-07
##                             IHWPvalue NormCounts_cond1_Mean
## TGCTCACTTTGGCAGCACATATACT 0.001663810                     0
## CTTCTCAGTCTTTTGGCTAAGATCA 0.002061792                     0
##                           NormCounts_cond1_Sd NormCounts_cond2_Mean
## TGCTCACTTTGGCAGCACATATACT                   0              6.599719
## CTTCTCAGTCTTTTGGCTAAGATCA                   0              6.000207
##                           NormCounts_cond2_Sd ClusterID Length BlastEvalue
## TGCTCACTTTGGCAGCACATATACT            1.821142       318     25    8.23e-09
## CTTCTCAGTCTTTTGGCTAAGATCA            1.707972       721     25    8.23e-09
##                           tRNA [Hh]sa ^U Other
## TGCTCACTTTGGCAGCACATATACT    0      0 26     0
## CTTCTCAGTCTTTTGGCTAAGATCA    0      0 17     0
##                                                                                                                                                                                                                                           FeatureList
## TGCTCACTTTGGCAGCACATATACT U6-L1687,U6-L1676,U6-L1561,U6-L1528,U6-L1493,U6-L1482,U6-L1306,U6-L1258,U6-L1248,U6-L1121,U6-L1110,U6-L1106,U6-L1093,U6-L1025,U6-L882,U6-L843,U6-L814,U6-L746,U6-L690,U6-L583,U6-L542,U6-L484,U6-L403,U6-L306,U6-L76,U6-L49
## CTTCTCAGTCTTTTGGCTAAGATCA                                                                                     U2-L1014,U2-L999,U2-L990,U2-L988,U2-L987,U2-L986,U2-L985,U2-L983,U2-L883,U2-L826,U2-L750,U2-L594,U2-L584,U2-L332,U2-L164,U2-L117,U2-L90</code></pre>
</div>
<div id="computing-the-na-fraction" class="section level3">
<h3 class="hasAnchor">
<a href="#computing-the-na-fraction" class="anchor"></a>Computing the NA fraction</h3>
<p>The function <a href="../reference/getNoBlastHitFraction.html">getNoBlastHitFraction</a> can be used to compute the fraction of reads without annotation for each sample. This fraction corresponds to the fraction of reads that are likely to be ignored in mapping approaches because they map to regions that are not annotated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/getNoBlastHitFraction.html">getNoBlastHitFraction</a></span>(summary, de_results<span class="op">$</span>norm_counts)</code></pre></div>
<pre><code>##          NA_fraction
## cond1_s1 0.016237479
## cond1_s2 0.017400664
## cond1_s3 0.017287479
## cond1_s4 0.015625000
## cond1_s5 0.017318969
## cond2_s1 0.005076142
## cond2_s2 0.006242367
## cond2_s3 0.006804610
## cond2_s4 0.005953174
## cond2_s5 0.006856780</code></pre>
<hr>
</div>
<div id="extended-expression-analysis-using-sequence-clusters" class="section level3">
<h3 class="hasAnchor">
<a href="#extended-expression-analysis-using-sequence-clusters" class="anchor"></a>Extended expression analysis using sequence clusters</h3>
<p>In addition to the default DEUS pipeline, we developed an adjusted approach that aggregates sequence counts for each sequence cluster and uses this information to find differentially expressed sequence clusters. By some small adjustments, the cluster expression analysis can be included into the default DEUS pipeline.</p>
<p>Instead of creating the sequence map for differentially expressed sequences only, it will be created on the raw input sequences since all sequences will be used for clustering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">map &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createMap.html">createMap</a></span>(count_table)</code></pre></div>
<p>Afterwards, all sequences can be clustered. Based on the clustering results, the sum of sequence counts will be created for each cluster via <a href="../reference/mergeAndAggregate.html">mergeAndAggregate</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cd_hit &lt;-<span class="st"> "/your/path/to/cdhit/cd-hit-est"</span>
all_seq_fasta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a></span>(count_table,map)
clust_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runClustering.html">runClustering</a></span>(cd_hit, all_seq_fasta, out_dir, <span class="dt">identity_cutoff=</span><span class="fl">0.9</span>, <span class="dt">length_cutoff=</span><span class="fl">0.9</span>, <span class="dt">wordlength=</span><span class="dv">9</span>, map)
cl_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeAndAggregate.html">mergeAndAggregate</a></span>(map, count_table, clust_result)</code></pre></div>
<p>For this manual, we will load the provided clustering results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clust_int &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"results/clust_result_clust_sequences.tsv"</span>, <span class="dt">package=</span><span class="st">"DEUS"</span>)
clust_result &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(clust_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(clust_result) &lt;-<span class="st"> </span>clust_result<span class="op">$</span>SequenceID
cl_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeAndAggregate.html">mergeAndAggregate</a></span>(map, count_table, clust_result)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(cl_counts, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##   cond1_s1 cond1_s2 cond1_s3 cond1_s4 cond1_s5 cond2_s1 cond2_s2 cond2_s3
## 0      142      119      152      146      139      148      112      150
## 1       91       96       75       84       85      120      129      108
##   cond2_s4 cond2_s5
## 0      167      136
## 1      128       96</code></pre>
<p>The resulting count matrix is then used as input for the DE analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="er">~</span><span class="st"> </span>condition
cl_de_results &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runDESeq2.html">runDESeq2</a></span>(cl_counts, pheno_info, design, <span class="dt">out_dir =</span> out_dir, <span class="dt">prefix =</span> <span class="st">"Cluster"</span>)
cl_sig_results &lt;-<span class="st"> </span>cl_de_results<span class="op">$</span>de_result
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(cl_sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##        Pvalue         padj  baseMean Log2FoldChange     lfcSE      stat
## 0 0.563340354 0.8110233879 133.23181     0.06532971 0.1130493 0.5778871
## 1 0.000151244 0.0006688245  95.71573     0.46504625 0.1227353 3.7890179
##     IHWPvalue
## 0 1.000000000
## 1 0.001355991</code></pre>
<p>Subsequently, the existing DEA results based on individual sequence counts and the generated DEA results based on sequence clusters can be merged into a single data frame via <a href="../reference/mergeSingleAndClusterResults.html">mergeSingleAndClusterResults</a>. Based on the individual and cluster Pvalues, the resulting data frame is filtered to detect sequences that are differentially expressed based on their sequence count or part of a differentially expressed sequence cluster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sig_results &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeSingleAndClusterResults.html">mergeSingleAndClusterResults</a></span>(cl_sig_results,clust_result,sig_results,map)
sig_results &lt;-<span class="st"> </span>sig_results[((<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(sig_results<span class="op">$</span>IHWPval) <span class="op">&amp;</span><span class="st"> </span>sig_results<span class="op">$</span>IHWPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>) <span class="op">|</span><span class="st"> </span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(sig_results<span class="op">$</span>Cl_IHWPval) <span class="op">&amp;</span><span class="st"> </span>sig_results<span class="op">$</span>Cl_IHWPval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)),]
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(sig_results, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                                 Pvalue       padj baseMean Log2FoldChange
## TCGCTTCTCAGCCTTTTGGCTAAGA 0.0007176335 0.03462413 7.263225      -1.521815
## CGCTTCTCAGCCTTTTGGCTAAGAT 0.0026808638 0.05501307 6.594336      -1.431923
##                               lfcSE      stat  IHWPvalue SequenceID
## TCGCTTCTCAGCCTTTTGGCTAAGA 0.4498748 -3.382752 0.04498802      seq_6
## CGCTTCTCAGCCTTTTGGCTAAGAT 0.4769671 -3.002143 0.04498802     seq_15
##                           Cl_Pvalue   Cl_padj Cl_baseMean
## TCGCTTCTCAGCCTTTTGGCTAAGA 0.2181573 0.4118952    56.53792
## CGCTTCTCAGCCTTTTGGCTAAGAT 0.2181573 0.4118952    56.53792
##                           Cl_Log2FoldChange Cl_lfcSE  Cl_stat Cl_IHWPvalue
## TCGCTTCTCAGCCTTTTGGCTAAGA          0.201767 0.163846 1.231443    0.6115901
## CGCTTCTCAGCCTTTTGGCTAAGAT          0.201767 0.163846 1.231443    0.6115901
##                           ClusterID
## TCGCTTCTCAGCCTTTTGGCTAAGA         2
## CGCTTCTCAGCCTTTTGGCTAAGAT         2</code></pre>
<p>Next, we run BLAST on all selected sequences.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_exec &lt;-<span class="st"> "/your/path/to/ncbi-blast-x.x.x/bin/blastn"</span>
blast_db &lt;-<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"blastdb/DASHR_subset.fa"</span>, <span class="dt">package =</span> <span class="st">"DEUS"</span>)
blast_ncores &lt;-<span class="st"> </span><span class="dv">2</span>
sig_seq_fasta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequencesAsFasta.html">sequencesAsFasta</a></span>(sig_results,map)
blast_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runBlast.html">runBlast</a></span>(blast_exec, blast_db, blast_ncores, sig_seq_fasta) </code></pre></div>
<p>In our example, we load pre-compiled data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">blast_int &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"results/blast_result_clust_sequences.tsv"</span>, <span class="dt">package=</span><span class="st">"DEUS"</span>)
blast_result &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(blast_int, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(blast_result)</code></pre></div>
<pre><code>##   SequenceID Annotation Length BlastEvalue
## 1      seq_5   U2-L1076     25    8.23e-09
## 2      seq_5   U2-L1072     25    8.23e-09
## 3      seq_5   U2-L1020     25    8.23e-09
## 4      seq_5   U2-L1019     25    8.23e-09
## 5      seq_5    U2-L974     25    8.23e-09
## 6      seq_5    U2-L973     25    8.23e-09</code></pre>
<p>Finally, the summary table is generated again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># merge results</span>
classes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"tRNA"</span>,<span class="st">"[Hh]sa"</span>,<span class="st">"^U"</span>)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeResults.html">mergeResults</a></span>(sig_results, count_stats, blast_result, clust_result, map)
summary &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCountsOfFeatureClasses.html">addCountsOfFeatureClasses</a></span>(summary, classes)
<span class="kw"><a href="../reference/writeSummaryFiles.html">writeSummaryFiles</a></span>(summary, out_dir)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(summary, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<pre><code>##                           SequenceID Log2FoldChange       Pvalue
## TGCTCACTTTGGCAGCACATATACT  seq_37001       4.575022 3.738245e-07
## CTTCTCAGTCTTTTGGCTAAGATCA  seq_37090       4.446684 9.264864e-07
##                             IHWPvalue Cl_Log2FoldChange    Cl_Pvalue
## TGCTCACTTTGGCAGCACATATACT 0.001663810          2.079899 3.094328e-13
## CTTCTCAGTCTTTTGGCTAAGATCA 0.002061792          3.820641 8.413101e-05
##                           Cl_IHWPvalue NormCounts_cond1_Mean
## TGCTCACTTTGGCAGCACATATACT 4.206725e-09                     0
## CTTCTCAGTCTTTTGGCTAAGATCA 2.063389e-03                     0
##                           NormCounts_cond1_Sd NormCounts_cond2_Mean
## TGCTCACTTTGGCAGCACATATACT                   0              6.599719
## CTTCTCAGTCTTTTGGCTAAGATCA                   0              6.000207
##                           NormCounts_cond2_Sd ClusterID Length BlastEvalue
## TGCTCACTTTGGCAGCACATATACT            1.821142      2493     25    8.23e-09
## CTTCTCAGTCTTTTGGCTAAGATCA            1.707972      7520     25    8.23e-09
##                           tRNA [Hh]sa ^U Other
## TGCTCACTTTGGCAGCACATATACT    0      0 26     0
## CTTCTCAGTCTTTTGGCTAAGATCA    0      0 17     0
##                                                                                                                                                                                                                                           FeatureList
## TGCTCACTTTGGCAGCACATATACT U6-L1687,U6-L1676,U6-L1561,U6-L1528,U6-L1493,U6-L1482,U6-L1306,U6-L1258,U6-L1248,U6-L1121,U6-L1110,U6-L1106,U6-L1093,U6-L1025,U6-L882,U6-L843,U6-L814,U6-L746,U6-L690,U6-L583,U6-L542,U6-L484,U6-L403,U6-L306,U6-L76,U6-L49
## CTTCTCAGTCTTTTGGCTAAGATCA                                                                                     U2-L1014,U2-L999,U2-L990,U2-L988,U2-L987,U2-L986,U2-L985,U2-L983,U2-L883,U2-L826,U2-L750,U2-L594,U2-L584,U2-L332,U2-L164,U2-L117,U2-L90</code></pre>
<p>In order to compare both aproaches, the function <a href="../reference/printClusterSummary.html">printClusterSummary</a> can be used to get detailed information about the number of significant sequences and clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/printClusterSummary.html">printClusterSummary</a></span>(summary)</code></pre></div>
<pre><code>## ## Sequence summary ##
## Significant sequences: 1906
## Significant sequences not clustered: 0
## Sequences in significant clusters: 22110
## Non-significant sequences in significant clusters: 20369
## - too low expressed: 20369
## - too weak differentially expressed: 0
## Significant sequences in non-significant clusters: 165
## ## Cluster summary ##
## Significant clusters: 4628
## Clusters including significant sequences: 1323
## Significant clusters including no significant sequences: 3429</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Tim Jeske, Maximilian Hastreiter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
